File: examples/go/core/shared/app_mcp_contracts.go
package shared

type AppMcpClientInfo struct {
	Name    string `json:"name"`
	Version string `json:"version"`
}

type AppMcpServerInfo struct {
	Name            string `json:"name"`
	Version         string `json:"version"`
	ProtocolVersion string `json:"protocolVersion"`
	Instructions    string `json:"instructions,omitempty"`
}

type AppMcpInitializeParams struct {
	ProtocolVersion string            `json:"protocolVersion"`
	Capabilities    Dict              `json:"capabilities,omitempty"`
	ClientInfo      *AppMcpClientInfo `json:"clientInfo,omitempty"`
}

type AppMcpInitializeResult struct {
	ProtocolVersion string `json:"protocolVersion"`
	Capabilities    struct {
		Tools struct {
			ListChanged bool `json:"listChanged"`
		} `json:"tools,omitempty"`
		Resources struct {
			ListChanged bool `json:"listChanged"`
		} `json:"resources,omitempty"`
	} `json:"capabilities"`
	ServerInfo struct {
		Name    string `json:"name"`
		Version string `json:"version"`
	} `json:"serverInfo"`
	Instructions string `json:"instructions,omitempty"`
}

type AppMcpTextContent struct {
	Type string `json:"type"`
	Text string `json:"text"`
}

type AppMcpToolDescriptor struct {
	Name         string    `json:"name"`
	Title        string    `json:"title,omitempty"`
	Description  string    `json:"description,omitempty"`
	InputSchema  AppSchema `json:"inputSchema"`
	OutputSchema AppSchema `json:"outputSchema,omitempty"`
	Annotations  Dict      `json:"annotations,omitempty"`
}

type AppMcpResourceDescriptor struct {
	URI         string `json:"uri"`
	Name        string `json:"name"`
	Title       string `json:"title,omitempty"`
	Description string `json:"description,omitempty"`
	MimeType    string `json:"mimeType,omitempty"`
	Annotations Dict   `json:"annotations,omitempty"`
}

type AppMcpTextResourceContent struct {
	URI      string `json:"uri"`
	MimeType string `json:"mimeType,omitempty"`
	Text     string `json:"text"`
}

type AppMcpCallResult struct {
	Content           []AppMcpTextContent `json:"content"`
	StructuredContent any                 `json:"structuredContent,omitempty"`
	IsError           bool                `json:"isError,omitempty"`
}

type AppMcpReadResourceResult struct {
	Contents []AppMcpTextResourceContent `json:"contents"`
}

type AppMcpRequestContext struct {
	Transport       string            `json:"transport"`
	RequestID       any               `json:"requestId,omitempty"`
	SessionID       string            `json:"sessionId,omitempty"`
	CorrelationID   string            `json:"correlationId,omitempty"`
	ClientInfo      *AppMcpClientInfo `json:"clientInfo,omitempty"`
	ProtocolVersion string            `json:"protocolVersion,omitempty"`
}

type AppMcpServer interface {
	ServerInfo() AppMcpServerInfo
	Initialize(params *AppMcpInitializeParams, parent *AppMcpRequestContext) (AppMcpInitializeResult, error)
	ListTools(parent *AppMcpRequestContext) ([]AppMcpToolDescriptor, error)
	ListResources(parent *AppMcpRequestContext) ([]AppMcpResourceDescriptor, error)
	ReadResource(uri string, parent *AppMcpRequestContext) (AppMcpReadResourceResult, error)
	CallTool(name string, args any, parent *AppMcpRequestContext) (AppMcpCallResult, error)
}

type BaseAppMcpAdapter interface {
	Transport() string
}

type BaseAppMcpProcessAdapter interface {
	BaseAppMcpAdapter
	Serve(server AppMcpServer) error
}

type AppMcpHTTPExchange struct {
	Method   string
	Path     string
	Headers  map[string]string
	BodyText string
}

type AppMcpHTTPResponse struct {
	StatusCode int
	Headers    map[string]string
	BodyText   string
}

type BaseAppMcpHTTPAdapter interface {
	BaseAppMcpAdapter
	EndpointPath() string
	Handle(exchange AppMcpHTTPExchange, server AppMcpServer) (*AppMcpHTTPResponse, error)
}

type AppMcpProtocolError struct {
	Code    int
	Message string
	Data    any
}

func (e *AppMcpProtocolError) Error() string {
	return e.Message
}

File: examples/kotlin/core/shared/app_mcp_contracts.kt
package app.protocol.examples.kotlin.core.shared

import app.protocol.examples.kotlin.core.AppSchema
import app.protocol.examples.kotlin.core.Dict

data class AppMcpClientInfo(
    val name: String,
    val version: String,
)

data class AppMcpServerInfo(
    val name: String,
    val version: String,
    val protocolVersion: String,
    val instructions: String? = null,
)

data class AppMcpInitializeParams(
    val protocolVersion: String,
    val capabilities: Dict<Any?>? = null,
    val clientInfo: AppMcpClientInfo? = null,
)

data class AppMcpInitializeResult(
    val protocolVersion: String,
    val capabilities: Capabilities,
    val serverInfo: ServerInfo,
    val instructions: String? = null,
) {
    data class Capabilities(
        val tools: ToolsCapability? = null,
        val resources: ResourcesCapability? = null,
    ) {
        data class ToolsCapability(
            val listChanged: Boolean? = null,
        )

        data class ResourcesCapability(
            val listChanged: Boolean? = null,
        )
    }

    data class ServerInfo(
        val name: String,
        val version: String,
    )
}

data class AppMcpTextContent(
    val type: String = "text",
    val text: String,
)

data class AppMcpToolDescriptor(
    val name: String,
    val title: String? = null,
    val description: String? = null,
    val inputSchema: AppSchema,
    val outputSchema: AppSchema? = null,
    val annotations: Dict<Any?>? = null,
)

data class AppMcpResourceDescriptor(
    val uri: String,
    val name: String,
    val title: String? = null,
    val description: String? = null,
    val mimeType: String? = null,
    val annotations: Dict<Any?>? = null,
)

data class AppMcpTextResourceContent(
    val uri: String,
    val mimeType: String? = null,
    val text: String,
)

data class AppMcpCallResult(
    val content: List<AppMcpTextContent>,
    val structuredContent: Any? = null,
    val isError: Boolean? = null,
)

data class AppMcpReadResourceResult(
    val contents: List<AppMcpTextResourceContent>,
)

data class AppMcpRequestContext(
    val transport: String,
    val requestId: Any? = null,
    val sessionId: String? = null,
    val correlationId: String? = null,
    val clientInfo: AppMcpClientInfo? = null,
    val protocolVersion: String? = null,
)

interface AppMcpServer {
    fun serverInfo(): AppMcpServerInfo

    suspend fun initialize(
        params: AppMcpInitializeParams? = null,
        parent: AppMcpRequestContext? = null,
    ): AppMcpInitializeResult

    suspend fun listTools(parent: AppMcpRequestContext? = null): List<AppMcpToolDescriptor>

    suspend fun listResources(parent: AppMcpRequestContext? = null): List<AppMcpResourceDescriptor>

    suspend fun readResource(
        uri: String,
        parent: AppMcpRequestContext? = null,
    ): AppMcpReadResourceResult

    suspend fun callTool(
        name: String,
        args: Any?,
        parent: AppMcpRequestContext? = null,
    ): AppMcpCallResult
}

abstract class BaseAppMcpAdapter {
    abstract val transport: String
}

abstract class BaseAppMcpProcessAdapter : BaseAppMcpAdapter() {
    abstract suspend fun serve(server: AppMcpServer)
}

data class AppMcpHttpExchange(
    val method: String,
    val path: String,
    val headers: Map<String, String?>,
    val bodyText: String? = null,
)

data class AppMcpHttpResponse(
    val statusCode: Int,
    val headers: Map<String, String> = emptyMap(),
    val bodyText: String? = null,
)

abstract class BaseAppMcpHttpAdapter : BaseAppMcpAdapter() {
    abstract val endpointPath: String

    abstract suspend fun handle(
        exchange: AppMcpHttpExchange,
        server: AppMcpServer,
    ): AppMcpHttpResponse?
}

class AppMcpProtocolError(
    val code: Int,
    override val message: String,
    val data: Any? = null,
) : Exception(message)
